File: microservices/booking/booking.py
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy
from flask_cors import CORS

app = Flask(__name__)  # making book.py as a Flask app
CORS(app)

app.config[
    'SQLALCHEMY_DATABASE_URI'] = 'mysql+mysqlconnector://root@localhost:3306/g7t3_booking'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
# @cross_origin(origin='*',headers=['Content-Type','Authorization'])
db = SQLAlchemy(app)


class Booking(db.Model):
    __tablename__ = 'booking'

    booking_id = db.Column(db.Integer, primary_key=True)
    customer_mobile = db.Column(db.String(8), nullable=False)
    provider_mobile = db.Column(db.String(8), nullable=False)
    provider_name = db.Column(db.String(128), nullable=True)
    booking_time = db.Column(db.String(5), nullable=False)
    booking_date = db.Column(db.String(10), nullable=False)
    booking_price = db.Column(db.Float(precision=2), nullable=False)
    booking_status = db.Column(db.Integer, nullable=False)
    booking_payment_status= db.Column(db.Integer, nullable=False)

    def __init__(self, booking_id, customer_mobile, provider_mobile, provider_name,
                 booking_time, booking_date, booking_price, booking_status, booking_payment_status):
        self.booking_id = booking_id
        self.customer_mobile = customer_mobile
        self.provider_mobile = provider_mobile
        self.provider_name = provider_name
        self.booking_time = booking_time
        self.booking_date = booking_date
        self.booking_price = booking_price
        self.booking_status = booking_status
        self.booking_payment_status = booking_payment_status

    def json(self):
        return {
            "booking_id": self.booking_id,
            "customer_mobile": self.customer_mobile,
            "provider_mobile": self.provider_mobile,
            "provider_name" : self.provider_name,
            "booking_time": self.booking_time,
            "booking_date": self.booking_date,
            "booking_price": self.booking_price,
            "booking_status": self.booking_status,
            "booking_payment_status": self.booking_payment_status
        }


@app.route("/booking")  # when will this page called? Default HTTP protocol is GET, even when not specified
def getAll():
    #return "Get all Books"  # pull the data from the DB
    return jsonify({
        "bookings": [booking.json() for booking in Booking.query.all()]
    })  #Book.query.all() is the same as " SELECT * FROM table_name "


@app.route("/booking/<string:customer_mobile>", methods=['GET'])
def findBooking(customer_mobile):
    allBookings = jsonify({
        "bookings": [
            booking.json() for booking in Booking.query.filter_by(
                customer_mobile=customer_mobile)
        ]
    })
    return allBookings

@app.route("/booking/provider/<string:provider_mobile>", methods=['GET'])
def findBookingByProvider(provider_mobile):
    allBookings = jsonify({
        "bookings": [
            booking.json() for booking in Booking.query.filter_by(
                provider_mobile=provider_mobile
            )
        ]
    })
    return allBookings

@app.route("/booking/<string:booking_id>", methods=['POST'])
def createBooking(booking_id):
    if (Booking.query.filter_by(booking_id=booking_id).first()):
        return jsonify({
            "message":
            "A booking with Booking ID '{}' already exists.".format(booking_id)
        }), 400

    data = request.get_json()
    print(data)
    booking = Booking(booking_id, **data)

    try:
        db.session.add(booking)
        db.session.commit()
    except:
        return jsonify({"message":
                        "An error occurred creating the booking."}), 500

    return jsonify(booking.json()), 201


@app.route("/booking/update/<string:booking_id>", methods=['POST'])
def updateCustomer(booking_id):
    if (not (Booking.query.filter_by(booking_id=booking_id).first())):
        return jsonify({
            "message":
            "A booking with Booking ID '{}' does not exists.".format(
                booking_id)
        }), 400

    data = request.get_json()
    booking = Booking(booking_id, **data)

    try:

        booking = Booking.query.filter_by(booking_id=booking_id).first()
        booking.booking_date = data['booking_date']
        booking.booking_time = data['booking_time']
        booking.customer_mobile = data['customer_mobile']
        booking.provider_mobile = data['provider_mobile']
        booking.booking_price = data['booking_price']
        db.session.commit()
    except:
        return jsonify({"message":
                        "An error occurred updating the booking."}), 500

    return jsonify(booking.json()), 201

@app.route("/booking/status/<string:booking_id>", methods=['POST'])
def updateBookingStatus(booking_id):
    if (not (Booking.query.filter_by(booking_id=booking_id).first())):
        return jsonify({
            "message":
            "A booking with Booking ID '{}' does not exists.".format(
                booking_id)
        }), 400
    try:
        booking = Booking.query.filter_by(booking_id = booking_id).first()
        booking.booking_status = 1
        db.session.commit()
    except:
        return jsonify({"message":"An error occurred updating the booking status"}), 500
    return jsonify(booking.json()), 201

@app.route("/booking/payment/<string:booking_id>", methods=['POST'])
def updatePaymentStatus(booking_id):
    if(not(Booking.query.filter_by(booking_id=booking_id).first())):
        return jsonify({
            "message":
            "A booking with Booking ID '{}' does not exists.".format(
                booking_id)
        }), 400
    try:
        booking = Booking.query.filter_by(booking_id = booking_id).first()
        booking.booking_payment_status = 1
        db.session.commit()
    except:
        return jsonify(
            {
                "message":"An error occurred updating the payment status"
            }
        ), 500
    return jsonify(booking.json()), 201

if __name__ == "__main__":  # to run this application with out having the name app.py
    app.debug = True
    app.run(host='0.0.0.0', port=1002, debug=True)

File: microservices/booking/booking.php
<?php
header("Access-Control-Allow-Origin: *");
session_start();
if (isset($_GET['mobile_number'])) {
    $customerMobile = $_GET['mobile_number'];
    $_SESSION['mobile_number'] = $customerMobile;
}

$HOST = "localhost";
$USERNAME = "root";
$PASSWORD = "";
$DB = "g7t3_serviceprovidertrial";
$link = mysqli_connect($HOST, $USERNAME, $PASSWORD, $DB, "3308");
if (!$link) {
    die(mysqli_error($link));
}
?>

<!DOCTYPE html>
<html>
    <head>
        <title>Bookings</title>
        <?php include("../../app/globalCSS.php"); ?>

        <style>
            .container-fluid {
                padding: 30px;
            }
        </style>

    <body>
        <?php include("../../app/globalCustomerHeader.php"); ?>
        <div class="container" style = "padding-top: 3%">
            <div class="row">
                <h1 class="display-4" style = "padding-bottom: 2%">Bookings</h1>          

                <table class="table" id="booking_table">
                    <thead>
                        <tr>
                            <th>Booking Number</th>
                            <th>Service Provider</th>
                            <th>Day</th>
                            <th>Time</th>
                            <th>Price</th>
                            <th>Review Status</th>
                            <th>Payment Status</th>
                        </tr>
                    </thead>
                </table>
            </div>

            <a id="addReviewBtn" class="btn btn-primary" href="createBooking.php" style = "margin-right: 1%">Create Booking</a>
            <a id="addReviewBtn" class="btn btn-primary" href="updateBooking.html">Update Booking</a>

        </div>
        
    </body>

    <script>
        function showError(message) {
            $("#booking_table").hide();
            $("#main_container").append("<label>" + message + "</label>");
        }

        $(async () => {
<?php
if (isset($_SESSION['mobile_number'])) {
    $customerMobile = $_SESSION['mobile_number'];
    ?>
                var customerMobile = <?php echo $customerMobile ?>;
                var serviceURL = "http://127.0.0.1:1002/booking/";
                serviceURL = serviceURL + customerMobile;
<?php } else {
    ?>
                var serviceURL = "http://127.0.0.1:1002/booking";
<?php }
?>
            console.log(serviceURL);
            try {
                const response = await fetch(serviceURL, {method: "GET"});
                const data = await response.json();
                console.log(data);
                var bookings = data.bookings;
                console.log(bookings);

                if (!bookings || !bookings.length) {
                    showError("No bookings found!");
                } else {
                    var rows = "";
                    for (const booking of bookings) {
                        var bookingDate = booking.booking_date;
                        bookingDate = bookingDate.split("-").reverse().join("/");
                        eachRow =
                                "<td>" + booking.booking_id + "</td>" +
                                "<td>" + booking.provider_name + "</td>" +
                                "<td>" + bookingDate + "</td>" +
                                "<td>" + booking.booking_time + "</td>" +
                                "<td> $" + booking.booking_price + "</td>";
                        if (booking.booking_status === 0) {
                            eachRow +=
                                    "<td>Service Not Yet Provided</td>";
                        } else {
                            var urlString = "<a href='../review/createReview.php?booking_id="
                                    + booking.booking_id + "'>";
                            eachRow += "<td>" +
                                    urlString +
                                    "Proceed Review" +
                                    "</a>" +
                                    "</td>";
                        }

                        if (booking.booking_payment_status === 0) {
                            eachRow +=
                            "<td>" + 
                                "<form action = 'makePayment.php' method = 'post'>" +
                                    "<input type='hidden' name='payment_id' value='" + booking.booking_id + "'/>" +
                                    "<input type='hidden' name='customer_mobile' value='<?php echo $customerMobile ?>'/>" +
                                    "<input type='hidden' name='provider_name' value='" + booking.provider_name + "'/>" + 
                                    "<input type='hidden' name='booking_price' value='" + booking.booking_price + "'/>" +
                                    "<input type='hidden' name='booking_id' value='" + booking.booking_id + "'/>" + 
                                    "<input type='hidden' name='service_provided' value='" + booking.booking_price + "'/>" +
                                    "<input type='submit' class = 'btn btn-primary'  value='Not Paid' />"
                                + "</form>" 
                            + "</td>"
                        
                        } else {
                            eachRow += "<td><button class = 'btn btn-danger disabled'>Paid</button></td>";
                        }
                        rows += "<tbody><tr>" + eachRow + "</tr></tbody>";
                    }
                    $('#booking_table').append(rows);
                }
            } catch (error) {
                // Errors when calling the service; such as network error, 
                // service offline, etc
                showError('There is a problem retrieving books data, please try again later.<br>' + error);
            } // error
        });
    </script>

</html>